trade-finance-manager-ui/server/controllers/case/cancellation/validation/validate-bank-request-date.test.ts

```diff
@@ -0,0 +1,90 @@
+    // Act
+    const day = date.getDate().toString();
+    const month = (date.getMonth() + 1).toString();
+    const year = date.getUTCFullYear().toString();
+
+    const result = validateBankRequestDate({ day, month, year });
+
+    // Assert
+    const expected: BankRequestDateValidationViewModel = {
+      errors: null,
+      bankRequestDate: date,
+    };
+    expect(result).toEqual(expected);
+  });
+
+  describe('custom date validation rules', () => {
+    it('returns an error if the bank request date is greater than 12 months in the future', () => {
+      const invalidBankRequestDate = add(new Date(), { months: 12, days: 1 });
+      const day = invalidBankRequestDate.getDate().toString();
+      const month = (invalidBankRequestDate.getMonth() + 1).toString();
+      const year = invalidBankRequestDate.getFullYear().toString();
+
+      const result = validateBankRequestDate({ day, month, year });
+
+      const expectedMessage = 'The bank request date cannot exceed 12 months in the future from the submission date';
+
+      const expectedError = {
+        summary: [{ text: expectedMessage, href: '#bank-request-date-day' }],
+        bankRequestDateError: { message: expectedMessage, fields: ['bank-request-date-day', 'bank-request-date-month', 'bank-request-date-year'] },
+      };
+
+      expect(result.errors).toEqual(expectedError);
+    });
+
+    it('returns an error if the bank request date is greater than 12 months in the past', () => {
+      const invalidBankRequestDate = add(new Date(), { months: -12, days: -1 });
+      const day = invalidBankRequestDate.getDate().toString();
+      const month = (invalidBankRequestDate.getMonth() + 1).toString();
+      const year = invalidBankRequestDate.getFullYear().toString();
+
+      const result = validateBankRequestDate({ day, month, year });
+
+      const expectedMessage = 'The bank request date cannot exceed 12 months in the past from the submission date';
+
+      const expectedError = {
+        summary: [{ text: expectedMessage, href: '#bank-request-date-day' }],
+        bankRequestDateError: { message: expectedMessage, fields: ['bank-request-date-day', 'bank-request-date-month', 'bank-request-date-year'] },
+      };
+
+      expect(result.errors).toEqual(expectedError);
+    });
+  });
+
+  describe('base date validation rules', () => {
+    it('should return the result of applyStandardValidationAndParseDateInput', () => {
+      const dateObjectToTest = { day: '', month: '1', year: '2024' };
+
+      const { errors: receivedErrors } = validateBankRequestDate(dateObjectToTest);
+
+      const { error: expectedError } = applyStandardValidationAndParseDateInput(dateObjectToTest, 'bank request date', 'bank-request-date');
+
+      expect(receivedErrors?.bankRequestDateError.message).toEqual(expectedError?.message);
+      expect(receivedErrors?.bankRequestDateError.fields).toEqual(expectedError?.fieldRefs);
```

Comment:
Could we just test the whole thing? e.g

```ts
const result = validateBankRequestDate(dateObjectToTest);

// ...

const expected = getErrorObjectFromMessageAndRefs();

expect(result).toEqual(expected);
```

Reply:
Ah the format we get from `applyStandardValidationAndParseDateInput` and `validateBankRequestDate` is different, so as you've mentioned there we'd have to call `getErrorObjectFromMessageAndRefs` in the test. I think it's best not to do that, as that function isn't exported, so it would be being exported just for a test here which I don't think is worth it.

Reply:
Ah, I think we should though @BethThomas141 - didn't notice that `getErrorObjectFromMessageAndRefs` isn't tested, can we add a test for this please

Then we can add a "full" test as mentioned above 😎

At the moment I could remove `getErrorObjectFromMessageAndRefs` and create something else and the test  would still pass

Reply:
I've exported `getErrorObjectFromMessageAndRefs` and added a test for it. 

I've just had a chat with @AlexBramhill and realised the best way to test this function as a unit test is to mock the response of `applyStandardValidationAndParseDateInput`. So I've now instead mocked this and check that the mocked response is returned. Have a look and let me know what you think! :)

Reply:
Thanks @BethThomas141 

Why do we need to mock `applyStandardValidationAndParseDateInput please?

Reply:
Can we update the `validateBankRequestDate` tests to use `getErrorObjectFromMessageAndRefs` and test the whole returned object please? So e.g instead of this:

```ts
const expectedError = {
  summary: [{ text: expectedMessage, href: '#bank-request-date-day' }],
  bankRequestDateError: { message: expectedMessage, fields: ['bank-request-date-day', 'bank-request-date-month', 'bank-request-date-year'] },
};

expect(result.errors).toEqual(expectedError);
```

We do this for example:

```ts
const expected = getErrorObjectFromMessageAndRefs('err message', fieldRefs);

expect(result).toEqual(expected);
```

Reply:
@ttbarnes `applyStandardValidationAndParseDateInput` should be mocked as that is a function call from outside the file - as this is a unit test we should be mocking returns like this. This unit test doesn't care what `applyStandardValidationAndParseDateInput` returns, just that it is called and we use it's result - what this returns is tested in the unit test for that function.